src/Neo/SmartContract/Native/Notary.cs

```diff
@@ -0,0 +1,333 @@
+        {
+            if (hardfork == ActiveIn)
+            {
+                engine.SnapshotCache.Add(CreateStorageKey(Prefix_MaxNotValidBeforeDelta), new StorageItem(DefaultMaxNotValidBeforeDelta));
+            }
+            return ContractTask.CompletedTask;
+        }
+
+        internal override async ContractTask OnPersistAsync(ApplicationEngine engine)
+        {
+            long nFees = 0;
+            ECPoint[] notaries = null;
+            foreach (Transaction tx in engine.PersistingBlock.Transactions)
+            {
+                var attr = tx.GetAttribute<NotaryAssisted>();
+                if (attr is not null)
+                {
+                    notaries ??= GetNotaryNodes(engine.SnapshotCache);
+                    var nKeys = attr.NKeys;
+                    nFees += (long)nKeys + 1;
+                    if (tx.Sender == Hash)
+                    {
+                        var payer = tx.Signers[1];
+                        var balance = engine.SnapshotCache.GetAndChange(CreateStorageKey(Prefix_Deposit, payer.Account))?.GetInteroperable<Deposit>();
+                        balance.Amount -= tx.SystemFee + tx.NetworkFee;
+                        if (balance.Amount.Sign == 0) RemoveDepositFor(engine.SnapshotCache, payer.Account);
+                    }
+                }
+            }
+            if (nFees == 0) return;
+            var singleReward = CalculateNotaryReward(engine.SnapshotCache, nFees, notaries.Length);
+
+            foreach (var notary in notaries) await GAS.Mint(engine, notary.EncodePoint(true).ToScriptHash(), singleReward, false);
+        }
+
+        protected override void OnManifestCompose(IsHardforkEnabledDelegate hfChecker, uint blockHeight, ContractManifest manifest)
+        {
+            manifest.SupportedStandards = new[] { "NEP-27" };
+        }
+
+        /// <summary>
+        /// Verify checks whether the transaction is signed by one of the notaries and
+        /// ensures whether deposited amount of GAS is enough to pay the actual sender's fee.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="sig">Signature</param>
+        /// <returns>Whether transaction is valid.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        private bool Verify(ApplicationEngine engine, byte[] sig)
+        {
+            Transaction tx = (Transaction)engine.ScriptContainer;
+            if (tx.GetAttribute<NotaryAssisted>() is null) return false;
+            foreach (var signer in tx.Signers)
+            {
+                if (signer.Account == Hash)
+                {
+                    if (signer.Scopes != WitnessScope.None) return false;
+                    break;
+                }
+            }
+            if (tx.Sender == Hash)
+            {
+                if (tx.Signers.Length != 2) return false;
+                var payer = tx.Signers[1].Account;
```

Comment:
Why don't get the payer from the attribute?

Reply:
Because attribute doesn't contain the payer. NotaryAssisted attribute contains only the number of signatures to be collected.